### PR TITLE
[netcore] Implement DependentHandle using the existing ephemeron code.

### PR DESCRIPTION
**Pull request by vargaz**

### CODE REVIEW

File: mcs/class/System.Private.CoreLib/System.Runtime.CompilerServices/DependentHandle.cs [hunk 1]
```diff
@@ -1,63 +1,67 @@
...
-	//        There's a handle allocated underneath. You must call Free() on this eventually
-	//        or you cause a native handle table leak.
+	// Instead of dependent handles, mono uses arrays of Ephemeron objects.
 	//
-	// This struct intentionally does no self-synchronization. It's up to the caller to
-	// to use DependentHandles in a thread-safe way.
-	//=========================================================================================
 	struct DependentHandle
 	{
 		IntPtr handle;
```
Comment (marek-safar) [inline mcs/class/System.Private.CoreLib/System.Runtime.CompilerServices/DependentHandle.cs]: This field could be removed

File: mcs/class/System.Private.CoreLib/System.Runtime.CompilerServices/DependentHandle.cs [hunk 2]
```diff
@@ -1,63 +1,67 @@
...
+			if (data [0].key == GC.EPHEMERON_TOMBSTONE)
+				return null;
+			return data [0].key;
+		}
 
-		public void SetPrimary (object primary) => throw new NotImplementedException ();
+		public object GetPrimaryAndSecondary (out object secondary) {
+			if (!IsAllocated)
+				throw new NotSupportedException ();
+			secondary = null;
```
Comment (marek-safar) [inline mcs/class/System.Private.CoreLib/System.Runtime.CompilerServices/DependentHandle.cs]: Could be set only for `if (data [0].key == GC.EPHEMERON_TOMBSTONE)`

File: mcs/class/System.Private.CoreLib/System.Runtime.CompilerServices/DependentHandle.cs [hunk 3]
```diff
@@ -1,63 +1,67 @@
...
-	//        No actual handle is allocated underneath. Illegal to call GetPrimary
-	//        or GetPrimaryAndSecondary(). Ok to call Free().
-	//
-	//        Initializing a DependentHandle using the nullary ctor creates a DependentHandle
-	//        that's in the !IsAllocated state.
-	//        (! Right now, we get this guarantee for free because (IntPtr)0 == NULL unmanaged handle.
-	//         ! If that assertion ever becomes false, we'll have to add an _isAllocated field
-	//         ! to compensate.)
-	//        
+	internal struct Ephemeron
```
Comment (marek-safar) [inline mcs/class/System.Private.CoreLib/System.Runtime.CompilerServices/DependentHandle.cs]: Could be `readonly struct`

**Comment by vargaz** [comment]
@monojenkins build failed

**Comment by vargaz** [comment]
@monojenkins build failed

**Pull request merged**